ledger/lruboxes.go

```diff
@@ -0,0 +1,132 @@
+
+import (
+	"github.com/algorand/go-algorand/logging"
+)
+
+//msgp:ignore cachedResourceData
+type cachedBoxData struct {
+	persistedBoxData
+
+	// box name
+	key string
+}
+
+// lruBoxes provides a storage class for the most recently used box data.
+// It doesn't have any synchronization primitive on it's own and require to be
+// syncronized by the caller.
+type lruBoxes struct {
+	// boxList contain the list of persistedBoxData, where the front ones are the most "fresh"
+	// and the ones on the back are the oldest.
+	boxList *persistedBoxesDataList
+
+	// boxes provides fast access to the various elements in the list by using the key
+	boxes map[string]*persistedBoxesDataListNode
+
+	// pendingBoxes are used as a way to avoid taking a write-lock. When the caller needs to "materialize" these,
+	// it would call flushPendingWrites and these would be merged into the boxes/boxesList
+	pendingBoxes chan cachedBoxData
+
+	// log interface; used for logging the threshold event.
+	log logging.Logger
+
+	// pendingWritesWarnThreshold is the threshold beyond we would write a warning for exceeding the number of pendingBoxes entries
+	pendingWritesWarnThreshold int
+}
+
+// init initializes the lruBoxes for use.
+// thread locking semantics : write lock
+func (m *lruBoxes) init(log logging.Logger, pendingWrites int, pendingWritesWarnThreshold int) {
+	m.boxList = newPersistedBoxList().allocateFreeNodes(pendingWrites)
+	m.boxes = make(map[string]*persistedBoxesDataListNode, pendingWrites)
+	m.pendingBoxes = make(chan cachedBoxData, pendingWrites)
+	m.log = log
+	m.pendingWritesWarnThreshold = pendingWritesWarnThreshold
+}
+
+// read the persistedBoxesData object that the lruBoxes has for the given key.
+// thread locking semantics : read lock
+func (m *lruBoxes) read(key string) (data persistedBoxData, has bool) {
+	if el := m.boxes[key]; el != nil {
+		return el.Value.persistedBoxData, true
+	}
+	return persistedBoxData{}, false
+}
+
+// flushPendingWrites flushes the pending writes to the main lruBoxes cache.
+// thread locking semantics : write lock
+func (m *lruBoxes) flushPendingWrites() {
+	pendingEntriesCount := len(m.pendingBoxes)
+	if pendingEntriesCount >= m.pendingWritesWarnThreshold {
+		m.log.Warnf("lruBoxes: number of entries in pendingBoxes(%d) exceed the warning threshold of %d", pendingEntriesCount, m.pendingWritesWarnThreshold)
+	}
+	for ; pendingEntriesCount > 0; pendingEntriesCount-- {
+		select {
+		case pendingBoxData := <-m.pendingBoxes:
+			m.write(pendingBoxData.persistedBoxData, pendingBoxData.key)
+		default:
+			return
+		}
+	}
+}
+
+// writePending write a single persistedBoxData entry to the pendingBoxes buffer.
+// the function doesn't block, and in case of a buffer overflow the entry would not be added.
+// thread locking semantics : no lock is required.
+func (m *lruBoxes) writePending(box persistedBoxData, key string) {
+	select {
+	case m.pendingBoxes <- cachedBoxData{persistedBoxData: box, key: key}:
+	default:
+	}
```

Comment:
question: if we write to a channel, is this the only way to do? or is there any special motivation behind such specific form of code? I wonder if JJ @jannotti knows anything tho...

Reply:
This is to avoid blocking, and drop the message instead if the channel buffer is full.
https://gobyexample.com/non-blocking-channel-operations